File: app/src/main/java/com/aaronbrecher/neverlate/viewmodels/CustomViewModelFactory.java
package com.aaronbrecher.neverlate.viewmodels;

import android.app.Application;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.aaronbrecher.neverlate.AppExecutors;
import com.aaronbrecher.neverlate.database.EventCompatibilityRepository;
import com.aaronbrecher.neverlate.database.EventsRepository;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class CustomViewModelFactory implements ViewModelProvider.Factory {

    private EventsRepository mEventsRepository;
    private Application mApplication;
    private AppExecutors mAppExecutors;

    @Inject
    public CustomViewModelFactory(EventsRepository eventsRepository, EventCompatibilityRepository compatabilityRepository,
                                  Application application, AppExecutors appExecutors) {
        this.mEventsRepository = eventsRepository;
        this.mApplication = application;
        this.mAppExecutors = appExecutors;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(MainActivityViewModel.class)) {
            return (T) new MainActivityViewModel(mEventsRepository,mApplication, mAppExecutors);
        } else if (modelClass.isAssignableFrom(DetailActivityViewModel.class)) {
            return (T) new DetailActivityViewModel(mEventsRepository, mAppExecutors);
        }
        else throw new IllegalArgumentException("ViewModel does not exist");
    }


}

File: app/src/main/java/com/aaronbrecher/neverlate/viewmodels/CustomViewModelFactory.kt
package com.aaronbrecher.neverlate.viewmodels

import android.app.Application
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.annotation.NonNull

import com.aaronbrecher.neverlate.AppExecutors
import com.aaronbrecher.neverlate.database.EventCompatibilityRepository
import com.aaronbrecher.neverlate.database.EventsRepository

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CustomViewModelFactory @Inject
constructor(private val mEventsRepository: EventsRepository, compatabilityRepository: EventCompatibilityRepository,
            private val mApplication: Application, private val mAppExecutors: AppExecutors) : ViewModelProvider.Factory {

    @NonNull
    override fun <T : ViewModel> create(@NonNull modelClass: Class<T>): T {
        return when {
            modelClass.isAssignableFrom(MainActivityViewModel::class.java) -> MainActivityViewModel(mEventsRepository, mApplication, mAppExecutors) as T
            modelClass.isAssignableFrom(DetailActivityViewModel::class.java) -> DetailActivityViewModel(mEventsRepository, mAppExecutors) as T
            else -> throw IllegalArgumentException("ViewModel does not exist")
        }
    }


}
